File: Zds/Zds.Flow/Updatables/SyncObject.vb
Imports Zds.Flow.Collections
Namespace Updatables
    <DebuggerStepThrough>
    Public Class SyncObject
        Inherits Updatable
        Implements ISyncObject
        Private _SyncUpdateEvent As SafeList(Of SyncObjectEventDelegate)
        Public Custom Event SyncUpdateEvent As SyncObjectEventDelegate
            AddHandler(value As SyncObjectEventDelegate)
                If _SyncUpdateEvent Is Nothing Then _SyncUpdateEvent = New SafeList(Of SyncObjectEventDelegate)
                _SyncUpdateEvent.Add(value)
            End AddHandler
            RemoveHandler(value As SyncObjectEventDelegate)
                If _SyncUpdateEvent IsNot Nothing Then _SyncUpdateEvent.Remove(value)
            End RemoveHandler
            RaiseEvent(Sender As ISyncObject)
                If _SyncUpdateEvent IsNot Nothing Then
                    For Each [Delegate] In _SyncUpdateEvent.Elements
                        [Delegate].Invoke(Sender)
                    Next
                End If
            End RaiseEvent
        End Event
        Private _IsLocked As Boolean
        Protected ReadOnly Property Lock As New Object
        Public ReadOnly Property IsLocked As Boolean Implements ISyncObject.IsLocked
            Get
                Return _IsLocked
            End Get
        End Property
        Protected Overrides Sub OnUpdated()
            MyBase.OnUpdated()
            If Not _IsLocked AndAlso Threading.Monitor.TryEnter(_Lock) Then
                _IsLocked = True
                Try
                    SyncUpdate()
                    _IsLocked = False
                    Threading.Monitor.Exit(_Lock)
                Catch ex As Exception
                    _IsLocked = False
                    Threading.Monitor.Exit(_Lock)
                    Handle(ex)
                End Try
            End If
        End Sub
        Protected Overridable Sub SyncUpdate() Implements ISyncObject.SyncUpdate
            RaiseEvent SyncUpdateEvent(Me)
        End Sub
        Sub New()
        End Sub
        Sub New(Updater As Updaters.IUpdater)
            MyBase.New(Updater)
        End Sub
        Sub New(SyncUpdate As SyncObjectEventDelegate)
            AddHandler SyncUpdateEvent, SyncUpdate
        End Sub
        Sub New(Updater As Updaters.IUpdater, SyncUpdate As SyncObjectEventDelegate)
            MyBase.New(Updater)
            AddHandler SyncUpdateEvent, SyncUpdate
        End Sub
        Public Delegate Sub SyncObjectEventDelegate(Sender As ISyncObject)
    End Class
End Namespace

File: Flow/Flow/Updatables/SyncObject.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using Flow.Interfaces;
using Flow.Updaters;

namespace Flow.Updatables
{
    public class SyncObject : Updatable
    {
        protected readonly object @lock = new object();
        public event SyncUpdateDelegate OnSyncUpdate;
        public bool IsLocked { get; private set; }
        protected override void InternalUpdate()
        {
            if (!IsLocked && Monitor.TryEnter(@lock))
            {
                IsLocked = true;
                try
                {
                    SyncUpdate();
                    IsLocked = false;
                    Monitor.Exit(@lock);
                }
                catch (Exception ex)
                {
                    IsLocked = false;
                    Monitor.Exit(@lock);
                    Handle(ex);
                }
            }
        }
        public override void Reset()
        {
            lock (actionLock)
            {
                Stop();
                Wait();
            }
        }
        public override bool Wait(TimeSpan timeout)
        {
            Stopwatch.Stopwatch stopwatch = new Stopwatch.Stopwatch();
            stopwatch.Start();
            while (IsLocked)
            {
                Thread.Sleep(1);
                if (!IsLocked) break;
                if (stopwatch.Elasped >= timeout)
                {
                    stopwatch.Stop();
                    return false;
                }
            }
            stopwatch.Stop();
            return true;
        }
        protected virtual void SyncUpdate() => OnSyncUpdate?.Invoke(this);

        public SyncObject() { }
        public SyncObject(IUpdater updater) : base(updater) { }

        public delegate void SyncUpdateDelegate(object sender);
    }

    public interface ISyncObject
    {
        bool IsLocked { get; }
    }
}
